src/coreComponents/linearAlgebra/utilities/ReverseCutHillMcKeeOrdering.cpp

```diff
@@ -0,0 +1,330 @@
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file ReverseCutHillMcKeeOrdering.hpp
+ */
+
+#include "ReverseCutHillMcKeeOrdering.hpp"
+
+namespace geos
+{
+
+void
+ReverseCutHillMcKeeOrdering::
+  computePermutation( localIndex const * const offsets,
+                      globalIndex const * const columns,
+                      localIndex const rankOffset,
+                      arrayView1d< localIndex > const perm )
+{
+  localIndex i = 0;
+  localIndex root = 0;
+  localIndex currentNum = 0;
+  localIndex const numRows = perm.size();
+
+  // get the degree for each node
+  array1d< localIndex > degree( numRows );
+  array1d< localIndex > marker( numRows );
+  for( i = 0; i < numRows; i++ )
+  {
+    degree[i] = offsets[i + 1] - offsets[i];
+    marker[i] = -1;
+  }
+
+  // start RCM loop
+  currentNum = 0;
+  while( currentNum < numRows )
+  {
+    // Find the unvisited node with the minimum degree
+    findNodeWithMinDegree( numRows, degree, marker, root );
+
+    // This is a new connected component
+    findPseudoPeripheralNode( offsets, columns, numRows, rankOffset, root, marker );
+
+    // Numbering of this component
+    computeNewOrdering( offsets, columns, numRows, rankOffset, root, marker, perm, currentNum );
+  }
+}
+
+void
+ReverseCutHillMcKeeOrdering::
+  findNodeWithMinDegree( localIndex const numRows,
+                         arrayView1d< localIndex > const degree,
+                         arrayView1d< localIndex > const marker,
+                         localIndex & rootRef )
+{
+  localIndex i = 0;
+  localIndex minDegree = numRows + 1;
+  localIndex root = 0;
+
+  for( i = 0; i < numRows; i++ )
+  {
+    if( marker[i] < 0 )
+    {
+      if( degree[i] < minDegree )
+      {
+        root = i;
+        minDegree = degree[i];
+      }
+    }
+  }
+  rootRef = root;
+}
+
+void
+ReverseCutHillMcKeeOrdering::
+  findPseudoPeripheralNode( localIndex const * const offsets,
+                            globalIndex const * const columns,
+                            localIndex const numRows,
+                            localIndex const rankOffset,
+                            localIndex & rootRef,
+                            arrayView1d< localIndex > const marker )
+{
+  localIndex i = 0;
+  localIndex r1 = 0;
+  localIndex r2 = 0;
+  localIndex row = 0;
+  localIndex minDegree = 0;
+  localIndex levDegree = 0;
+  localIndex numLevels = 0;
+  localIndex newNumLevels = 0;
+  localIndex root = rootRef;
+
+  // at most numRows levels
+  array1d< localIndex > level_i( numRows+1 );
+  array1d< localIndex > level_j( numRows );
```

Comment:
I'm not sure if these 2 are allocations. But if they are, I would allocate once in `computePermutation` and pass those arrays as parameters for each call to `findPseudoPeripheralNode`. Maybe there's a threading aspect I'm missing.